Sync Site Details when the siteurl/home changes

andrea-sdl opened the pull request:
## Description
This PR improves the SDS syncing by adding a sync when the `siteurl/home changes`
This change should support both single and multisites.

## Changelog Description
### Improved Site Details sync

Small improvements in the site details sync to trigger a sync when the home/siteurl changes.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [X] This change works and has been tested locally (or has an appropriate fallback).
- [X] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [X] I've created a changelog description that aligns with the provided examples.

## Steps to Test

*Single site*

1. Update the home/siteurl of a site `wp option update siteurl 'http://example.com' && wp option update home 'http://example.com'` 
1. Check that the SDS endpoint has been call and that the data has changed.

*Network site*

1. Update the home/siteurl of a site `wp option update siteurl 'http://example.com' --url=<your-old-url> && wp option update home 'http://example.com'  --url=<your-old-url>` 
1. Check that the SDS endpoint has been call and that the data has changed.

andrea-sdl committed:
added support for updating site details on update siteurl

andrea-sdl committed:
added fastcgi_finish_request to flush content and renamed functions

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

sjinks committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

rebeccahum committed:
Revert https://github.com/Automattic/vip-go-mu-plugins/pull/3899 (#3931)

rebeccahum committed:
Search: Update breakpoint time for stale indices (#3932)

rebeccahum committed:
PR Template: Add roll out to checklist (#3933)

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
introduced a method to expose the should_run_sds_sync variable

andrea-sdl committed:
added tests

dependabot[bot] committed:
chore(deps-dev): bump lint-staged from 13.0.3 to 13.0.4 (#3938)

Bumps [lint-staged](https://github.com/okonet/lint-staged) from 13.0.3 to 13.0.4.
- [Release notes](https://github.com/okonet/lint-staged/releases)
- [Commits](https://github.com/okonet/lint-staged/compare/v13.0.3...v13.0.4)

---
updated-dependencies:
- dependency-name: lint-staged
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps-dev): bump playwright from 1.28.0 to 1.28.1 in /__tests__/e2e (#3937)

Bumps [playwright](https://github.com/Microsoft/playwright) from 1.28.0 to 1.28.1.
- [Release notes](https://github.com/Microsoft/playwright/releases)
- [Commits](https://github.com/Microsoft/playwright/compare/v1.28.0...v1.28.1)

---
updated-dependencies:
- dependency-name: playwright
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps-dev): bump @playwright/test in /__tests__/e2e (#3935)

Bumps [@playwright/test](https://github.com/Microsoft/playwright) from 1.28.0 to 1.28.1.
- [Release notes](https://github.com/Microsoft/playwright/releases)
- [Commits](https://github.com/Microsoft/playwright/compare/v1.28.0...v1.28.1)

---
updated-dependencies:
- dependency-name: "@playwright/test"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps-dev): bump eslint from 8.27.0 to 8.28.0 in /__tests__/e2e (#3936)

Bumps [eslint](https://github.com/eslint/eslint) from 8.27.0 to 8.28.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/main/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v8.27.0...v8.28.0)

---
updated-dependencies:
- dependency-name: eslint
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Tried to support changes done through the network admin

andrea-sdl committed:
Merge remote-tracking branch 'origin/add/PIE-3294-sds-sync' into add/PIE-3294-sds-sync

andrea-sdl committed:
Revert "Tried to support changes done through the network admin"

This reverts commit 53248f4e221de53f89e37839f1edb1cbb841b9df.

andrea-sdl committed:
exclude network admin from the changes

andrea-sdl committed:
delay network-admin changes in a specific cron.

andrea-sdl committed:
add support for both instant sync on shutdown and quick sync via cron

andrea-sdl committed:
changed tests to support new blogs_to_sync field

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

rinatkhaziev commented on config/class-sync.php:
```diff
@@ -63,7 +77,43 @@ public function filter_cron_schedules( $schedule ) {
+		}
+
+		if ( false !== array_search( $original_blog_id, $this->blogs_to_sync ) ) {
+			if ( function_exists( 'fastcgi_finish_request' ) ) {
```
This block should be moved to the top of the method because everything below is expensive too (especially switch_to_blog/restore_current_blog)

andrea-sdl replied:
@rinatkhaziev moved up in https://github.com/Automattic/vip-go-mu-plugins/pull/3742/commits/9af15e80e1088b8aaf620612ab6a437937ce82ea 
I also like the idea that we run this _before_ the `switch_to_blog`.

rinatkhaziev commented on config/class-sync.php:
```diff
@@ -12,6 +12,11 @@ class Sync {
+	/**
+	 * @var array List of blogIDs that need a sync
+	 */
+	private $blogs_to_sync = [];
```
If we use an associative array `[ $blog_id => $blog_id,  $blog_2_id => $blog_2_id ]`  we can reduce check's complexity to O(1): 
instead of `array_search` or `is_array` this becomes a simple `isset( $this->blog_ids_to_sync[ get_current_blog_id() ] )`

WPprodigy replied:
This list should never be big, in fact I'm not sure that it will ever be more than 1 - at least through core functionality. It's either the blog you are on, or another blog being updated in network admin. Pretty sure the only way it grows is through custom code that loops through blogs and does an update.

Even in that case though, with 10k entries we're still talking milliseconds. I'd argue if it's worth the complexity & memory difference. But also, neither is large enough to matter either 😆

andrea-sdl replied:
@rinatkhaziev I don't have strong opinions on either implementation. Is this a blocker for you?

rinatkhaziev replied:
yikes, missed I this comment, sorry! @andrea-sdl

rinatkhaziev reviewed (changes requested):
Thanks, looking mostly good, but left some inline feedback

andrea-sdl committed:
moved the put_site_details code up so it's run before the switch_to_blog

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

rinatkhaziev committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
introduced the BLOGS_TO_SYNC_LIMIT to limit the amount of blogs we queue to sync.
This reduces the possibility of abuses, we'll rely on the cron for the rare cases where we need to sync a bigger batch.

WPprodigy committed:
update comments

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
introduce the is_sync_ratelimited to limit the number of syncs we do in a specific interval

andrea-sdl committed:
save original_blog_id so that we know it hasn't been altered after the init.

andrea-sdl committed:
Queue the sync only if we're in the admin/WP CLI context, to reduce frontend impact

andrea-sdl committed:
replace manual WP_CLI check with Context::is_wp_cli

andrea-sdl committed:
Run the sync checks only if we have blogs to sync in the array

andrea-sdl committed:
use the correct option name (was homeurl, but the option is home)

andrea-sdl committed:
Add an extra test to check the logic that limits the blogs sync to a maximum of 10

andrea-sdl commented on config/class-sync.php:
```diff
@@ -22,17 +32,27 @@ public static function instance() {
-		$is_cli = defined( 'WP_CLI' ) && WP_CLI;
-
-		if ( ! is_admin() && ! $is_cli ) {
+		if ( ! is_admin() && ! Context::is_wp_cli() ) {
```
@WPprodigy do you see any problems moving the `$is_cli` to the `Context::is_wp_cli()`?

WPprodigy replied:
Nope, I like the change :)

WPprodigy replied:
Nope, I like the change!

andrea-sdl commented on config/class-sync.php:
```diff
@@ -63,7 +83,65 @@ public function filter_cron_schedules( $schedule ) {
+
+	public function queue_sync_for_blog() {
+		// Queue the sync only from admin or CLI, to keep it out of frontend request path
+		if ( ! is_admin() && ! Context::is_wp_cli() ) {
+			return;
+		}
```
@WPprodigy @rinatkhaziev one thing I've also added is to not queue the sync if we're not on admin/WP CLI, like we already do for the `maybe_setup_cron`. 

I didn't include this into the `init_listener` primarily to make it easier to test (I was having troubles defining the test to check that the behavior would actually change if we're not on WP CLI)

WPprodigy replied:
I like the restriction as it brings about some extra safety.

There's a (small) chance the option is updated via REST/XMLRPC endpoints, which we wouldn't catch here. Maybe that's an okay trade-off though. (Like you said earlier, we can always rely on the normal cron schedule to catch)

Also, `is_admin()` does unfortunately include frontend (possibly unauthenticated) AJAX requests. We do need to support authenticated admin ajax requests though since that is a likely route used for updating options in the admin.

andrea-sdl commented on config/class-sync.php:
```diff
@@ -63,7 +83,65 @@ public function filter_cron_schedules( $schedule ) {
+			$needs_sync = true;
+		}
+
+		if ( $needs_sync && ! $this->is_sync_ratelimited() ) {
```
rate-limiting the sync ( @WPprodigy suggestion )

andrea-sdl commented on config/class-sync.php:
```diff
@@ -63,7 +83,65 @@ public function filter_cron_schedules( $schedule ) {
+
+		// To avoid performance issues, don't add if the count would surpass BLOGS_TO_SYNC_LIMIT.
+		// Can rely on the default cron schedules to sync the rest.
+		if ( count( $this->blogs_to_sync ) >= self::BLOGS_TO_SYNC_LIMIT ) {
```
We will also limit the number of blogs we sync via a cron trigger, to limit inadvertently triggering a mass update on SDS that might cause problems.

We will rely on our cronjobs for when there is a mass update.

andrea-sdl commented on config/class-sync.php:
```diff
@@ -22,17 +32,27 @@ public static function instance() {
 
 	public function init() {
+		// Saving the initial blog_id in the init to be extra sure nobody changed it later on via a switch_to_blog().
+		$this->original_blog_id = get_current_blog_id();
```
@WPprodigy I'm not sure if placing this here is early enough or if we should add it even before the init, to make sure we are getting the "first blog_id". Ideas?

WPprodigy replied:
This should be good, happens on the `init` WP hook right now which is quite early. If we want to be really safe, could make sure it can't be re-set (in case some code triggers the hook again). So something like this perhaps would be good:

```
if ( is_null( $this->original_blog_id ) ) {
	$this->original_blog_id = get_current_blog_id();
}
```

andrea-sdl replied:
Change added in https://github.com/Automattic/vip-go-mu-plugins/pull/3742/commits/d1755d951d6421890ff5fd06801b4ad93a8ac4a6

andrea-sdl committed:
fix test-sync test so they ar run in separate processes (aside from the multisite one)

This is to avoid failure on single site tests, which required the @runInSeparateProcess, which is not applicable here because of https://core.trac.wordpress.org/ticket/51773

WPprodigy commented on config/class-sync.php:
```diff
@@ -63,7 +83,65 @@ public function filter_cron_schedules( $schedule ) {
+	}
+
+	public function run_sync_checks() {
+		//don't even try to run this if we don't have blogs to sync.
+		if ( count( $this->blogs_to_sync ) == 0 ) {
+			return;
+		}
+		$needs_sync = false;
+
+		// Check if the current request changed important data on this blog.
+		$blog_had_changes = false !== array_search( $this->original_blog_id, $this->blogs_to_sync );
+		if ( $blog_had_changes ) {
+			$needs_sync = true;
+		}
```
```suggestion
		if ( 0 === count( $this->blogs_to_sync ) ) {
			return;
		}

		$needs_sync = false !== array_search( $this->original_blog_id, $this->blogs_to_sync );
```

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Update config/class-sync.php

Co-authored-by: Caleb Burks <caleb.burks@automattic.com>

andrea-sdl committed:
added an extra null check to avoid overriding the original_blog_id in case the init gets called more than once.

andrea-sdl committed:
clearing the $this->blogs_to_sync queue to ensure we don't sync the same blog twice in case the function is called more than once

andrea-sdl commented on config/class-sync.php:
```diff
@@ -63,7 +85,61 @@ public function filter_cron_schedules( $schedule ) {
+			}
+		}
+		//clearing the sync queue after the run so that we don't sync the same blog twice
+		$this->blogs_to_sync = array();
```
@WPprodigy I've also added this to clear the sync queue after we're done with it. If there is a slight possibility that the function might get triggered more than once then I want to be sure it's not sending double the notifications it needs.

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

andrea-sdl committed:
Merge branch 'develop' into add/PIE-3294-sds-sync

Commit:
Regenerate the bundle

WPprodigy committed:
Reduce max syncs per interval to 10

WPprodigy committed:
Revert "Regenerate the bundle"

This reverts commit 453a1cebc2a7d6fb5a6bf06d9b96c7f4aa8feb84.

WPprodigy reviewed (dismissed):
Did some testing, looking good!

Commit:
Regenerate the bundle

WPprodigy committed:
Merge branch 'develop' into add/PIE-3294-sds-sync
